### PR TITLE
[Property wrappers] Prototype support for enclosing self-based access

### PR DESCRIPTION
**Pull request by DougGregor**
Allow property wrapper types to support a second access pattern for
instance properties of classes. When supported, the property wrapper's
static subscript(_enclosingInstance:storage:) is provided with the
enclosing "self" and a reference-writable key path referring to the
backing storage property.

Implements rdar://problem/52222560.

### CODE REVIEW

**Comment by DougGregor** [comment]
@swift-ci please smoke test

**Comment by DougGregor** [comment]
@swift-ci please build toolchain

**Comment by DougGregor** [comment]
@swift-ci please smoke test

**Comment by DougGregor** [comment]
@swift-ci please smoke test

**Comment by DougGregor** [comment]
@swift-ci please build toolchain

**Comment by DougGregor** [comment]
@swift-ci please build toolchain

File: include/swift/AST/DiagnosticsSema.def [hunk 1]
```diff
@@ -4452,6 +4452,9 @@ ERROR(property_wrapper_type_requirement_not_accessible,none,
       "more restrictive access than its enclosing property wrapper type %3 "
       "(which is %select{private|fileprivate|internal|public|open}4)",
       (AccessLevel, DescriptiveDeclKind, DeclName, Type, AccessLevel))
+ERROR(property_wrapper_ambiguous_enclosing_self_subscript, none,
+      "property wrapper type %0 has multiple enclosing-self subscripts %1",
+      (Type, DeclName))
```
Comment (jckarter) [inline include/swift/AST/DiagnosticsSema.def]: Does it work if you have something like this, to handle a value-semantics wrapper type?

```
// Read only access from a read only base
subscript(_enclosingInstance: EnclosingSelf, storage: KeyPath<EnclosingSelf, Self>) -> Wrapped { get }

// Writable projection from a writable base
subscript(_enclosingInstance: inout EnclosingSelf, storage: WritableKeyPath<EnclosingSelf, Self>) -> Wrapped { get set }
```

File: lib/Sema/CodeSynthesis.cpp [hunk 2]
```diff
@@ -613,6 +616,57 @@ namespace {
   };
 } // end anonymous namespace
 
+namespace  {
+  /// Describes the information needed to perform property wrapper access via
+  /// the enclosing self.
+  struct EnclosingSelfPropertyWrapperAccess {
+    /// The (genreric) subscript that will be used to perform the access.
```
Comment (jckarter) [inline lib/Sema/CodeSynthesis.cpp]: typo

**Comment by jckarter** [comment]
Any tests?

**Comment by lattner** [comment]
This appears to be a pretty substantial expansion of the property wrappers feature, which isn't yet standardized.  It seems like a really nice extension, but are you planning to go through swift-evolution with this?

**Comment by andyj-at-aspin** [comment]
Um. Is there a problem with the Swift toolchain in Xcode 11 GM?

I have replicated the code from the Unit Tests in this pull request into a test project. As coded, with the declaration of the Observable property wrapper and MyType test class in the same source file it all works spiffingly but the moment I mode the definition of the MyType class to another source file, mimicking definition the property wrapper in a utility file and making use of it in many others...

Xcode spits out a Segmentation fault 11 and will not compile.

If you can only use property wrappers with the so useful enclosing self subscript creation in the file in which you define them then that's going to mean a return to boiler plate for me. I was sooooo looking forward to using this to finally create the thread safe lazy that I've been boiler plating around for the last 3 years!

[SegFault.txt](https://github.com/apple/swift/files/3601888/SegFault.txt)

**Comment by jckarter** [comment]
@andyj-at-aspin Please file a bug at bugs.swift.org, if you haven't already. Thanks for reporting this!

**Pull request merged**